File: vol1/ch2/LeetCode1634.py
# Definition for polynomial singly-linked list.
# class PolyNode:
#     def __init__(self, x=0, y=0, next=None):
#         self.coefficient = x
#         self.power = y
#         self.next = next

class Solution:
    def addPoly(self, poly1: 'PolyNode', poly2: 'PolyNode') -> 'PolyNode':
        p,q=poly1,poly2
        if p==None: return q
        if q==None: return p
        h=PolyNode(); r=h
        while p!=None and q!=None:
            if p.power>q.power:
                r.next=p; r=p
                p=p.next
            elif q.power>p.power:
                r.next=q; r=q;
                q=q.next
            else:
                c=p.coefficient+q.coefficient
                if c!=0:
                    p.coefficient=c;
                    r.next=p;
                    r=p;
                p,q=p.next,q.next
        if p!=None: r.next=p
        else: r.next=q
        return h.next 

File: vol1/ch2/LeetCode1634.cpp
class Solution {
public:
    PolyNode* addPoly(PolyNode* poly1, PolyNode* poly2) {
        PolyNode*p=poly1;
        PolyNode*q=poly2;
        if(p==NULL) return q;
        if(q==NULL) return p;
        PolyNode*h=new PolyNode();
        PolyNode*r=h,*s;
        while(p!=NULL && q!=NULL) {
            if(p->power>q->power) {
                s=new PolyNode(p->coefficient,p->power);
                r->next=s;
                r=s;
                p=p->next;
            }
            else if(q->power>p->power) {
                s=new PolyNode(q->coefficient,q->power);
                r->next=s;
                r=s;
                q=q->next;
            }
            else {
                int c=p->coefficient+q->coefficient;
                if(c!=0) {
                    s=new PolyNode(c,p->power);
                    r->next=s;
                    r=s;
                }
                p=p->next;
                q=q->next;
            }
        }
        if(q!=NULL) p=q;
        while(p!=NULL) {
            s=new PolyNode(p->coefficient,p->power);
            r->next=s;
            r=s;
            p=p->next;
        }
        return h->next;
    }
};

